lib/TPP/Transforms/TileConfig.cpp

```diff
@@ -0,0 +1,154 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements loop tiling on parallel loops.
+//
+//===----------------------------------------------------------------------===//
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Transforms/Utils/VNNIUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include <iostream>
+#include <list>
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_TILECONFIGINSERTIONPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+
+namespace mlir {
+namespace tpp {
+
+template <typename InvokeOpTy, typename DispatchOpTy>
+static void appendBrgemmFlags(SmallVector<Attribute> &attributes,
+                              PatternRewriter &rewriter, InvokeOpTy opTy) {
+  auto flags =
+      dyn_cast<DispatchOpTy>(opTy.getOperand(0).getDefiningOp()).getFlags();
+  for (auto flagItr : flags) {
+    if (flagItr == xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                            xsmm::GemmFlags::NONE)) {
+      return;
+    }
+    attributes.push_back(flagItr);
+  }
+
+  if (attributes.empty())
+    attributes.push_back(
+        xsmm::GemmFlagsAttr::get(rewriter.getContext(), xsmm::GemmFlags::NONE));
+}
+
+template <typename InvokeOpTy, typename DispatchOpTy>
+struct TileConfig : OpRewritePattern<InvokeOpTy> {
+  using OpRewritePattern<InvokeOpTy>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(InvokeOpTy op,
+                                PatternRewriter &rewriter) const override {
+    if (xsmm::utils::getDataType(rewriter, op.getOperand(1).getType()) !=
+        xsmm::DataTypeAttr::get(rewriter.getContext(), xsmm::DataType::BF16))
+      return failure();
+    auto flags =
+        dyn_cast<DispatchOpTy>(op.getOperand(0).getDefiningOp()).getFlags();
+    for (auto flagItr : flags) {
+      if (flagItr == xsmm::GemmFlagsAttr::get(
+                         rewriter.getContext(),
+                         mlir::xsmm::GemmFlags::NO_RESET_TILECONFIG) ||
+          flagItr == xsmm::GemmFlagsAttr::get(
+                         rewriter.getContext(),
+                         mlir::xsmm::GemmFlags::NO_SETUP_TILECONFIG)) {
```

Comment:
Do these flag have some meaning for xsmm calls or is the only use for them just pattern convergence?
If it's the latter, then op annotation would be simpler. Although, in this case perhaps `PatternRewriter` is not the best tool overall but the pass could just once walk over all xsmm ops in a function.

Reply:
It's a bit convoluted but they do. 

Basically, for correctness issues, *every* GEMM-like op carries the flags `SETUP` and `RESET`. Otherwise, you run the risk of running an uninitialized AMX instruction. 

These ones are saying "don't setup" (ie. `RESET`) or "don't reset" (ie. `SETUP`), and they're executed before/after the outer loops (OMP thread). The inner loop's GEMM now has *both* `NO SETUP` and `NO RESET`, to make sure it does not incurr in the penalty of setting up AMX unnecessarily. This is the heart of the optimization.

The current code in the PR is hoping that LICM will hoist the setup/reset invokes into the outer parallel loop, but it doesn't, because it's not side-effect free (there's an alloca in there). So we agreed to move the ops in the right place (between `scf.parallel` and the first `scf.for`) by hand for now.

As why they're flags, because libxsmm was modified to work that way. May not be the best API but it is what it is now. We can change that later (and I'd prefer a simpler API too).

Reply:
This API is breaking the ABI. Hence it's negated. 

I still think it's a bug / too conservative that LICM is not moving the stuff. the alloca has nothing to do with the loops nest. We can proof that the memory we are allocating/accessing has nothing to do with induction variables and that the call we are making doesn't consider the induction variables.